File: src/main/java/me/stavgordeev/plugin/commands/BlueprintBazaarCommands.java
package me.stavgordeev.plugin.commands;

import me.stavgordeev.plugin.Minigames.BlueprintBazaar;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class BlueprintBazaarCommands extends MinigameCommandsSkeleton {
    private final BlueprintBazaar blueprintbazaar;
    public BlueprintBazaarCommands(BlueprintBazaar blueprintbazaar) {
        this.blueprintbazaar = blueprintbazaar;
    }

    @Override
    protected boolean handleCommand(Player player, Command command, String label, String[] args) {
        switch (args[0].toLowerCase()) {
            case "start":
                try {
                    blueprintbazaar.start(player);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
                break;
            case "start_hard_mode":
                try {
                    blueprintbazaar.startFastMode(player);
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
                break;
            case "stop":
                blueprintbazaar.pauseGame(player);
                break;
            case "resume":
                blueprintbazaar.resumeGame(player);
                break;
            case "end":
                blueprintbazaar.endGame(player);
                break;
            case "nuke_area":
                blueprintbazaar.nukeArea(player.getLocation(), 50);
                break;
            default:
                Bukkit.getServer().broadcast(Component.text("Unknown command.").color(NamedTextColor.RED));
                break;
        }
        return true;
    }

    @Override
    protected @Nullable List<String> handleTabComplete(CommandSender sender, Command command, String label, String[] args) {
        if (args.length == 1) {
            return List.of("start", "stop", "start_hard_mode", "resume", "end", "nuke_area");
        }
        return List.of();
    }
}

File: src/main/java/base/minigames/blueprint_bazaar/BlueprintBazaarCommands.kt
package base.minigames.blueprint_bazaar

import base.commands.MinigameCommandsSkeleton
import base.minigames.maze_hunt.MazeHuntCommands
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class BlueprintBazaarCommands(private val blueprintbazaar: BlueprintBazaar) : MinigameCommandsSkeleton() {
        /**
     * All sub-commands for this minigame
     */
    private enum class SubCommands {
        START,
        START_HARD_MODE,
        PAUSE,
        RESUME,
        END,
        NUKE_ARENA,
        SPAWN_BUILD,
        SHOWCASE_ALL_BUILDS,
        INIT_SCHEMATICS,
        CYCLE_THROUGH_SCHEMATICS
        ;

        companion object {
            /**
             * Converts a string to a SubCommand enum value. Case-insensitive.
             * @param str The string to convert
             * @return The SubCommand enum value, or null if the string does not match any enum value
             */
            fun fromString(str: String): SubCommands? {
                return entries.find { it.name.equals(str, ignoreCase = true) }
            }
        }
    }



    override fun handleCommand(sender: Player, command: Command, label: String, args: Array<String>): Boolean {
        when (SubCommands.fromString(args[0])) {
            SubCommands.START -> {
                if (blueprintbazaar.stopIfGameIsRunning()) return false
                blueprintbazaar.start(sender)
            }
            SubCommands.START_HARD_MODE -> {
                if (blueprintbazaar.stopIfGameIsRunning()) return false
                blueprintbazaar.startFastMode(sender)
            }
            SubCommands.PAUSE -> {
                if (blueprintbazaar.stopIfGameIsPaused()) return false
                blueprintbazaar.pauseGame()
            }
            SubCommands.RESUME -> {
                if (blueprintbazaar.stopIfGameIsNotPaused()) return false
                blueprintbazaar.resumeGame()
            }
            SubCommands.END -> {
                if (blueprintbazaar.stopIfGameIsNotRunning()) return false
                blueprintbazaar.endGame()
            }
            SubCommands.INIT_SCHEMATICS -> blueprintbazaar.initSchematics()
            SubCommands.SPAWN_BUILD -> blueprintbazaar.skipToNextBuild()
            SubCommands.SHOWCASE_ALL_BUILDS -> blueprintbazaar.loadAllSchematics()
            SubCommands.CYCLE_THROUGH_SCHEMATICS -> blueprintbazaar.cycleThroughSchematics()
            SubCommands.NUKE_ARENA -> blueprintbazaar.nukeArea(BPBConst.Locations.GAME_START_LOCATION, BPBConst.Locations.GAME_AREA_RADIUS)

            else -> return error(sender, "Unknown command.")
        }

        return true
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>
    ): List<String> {
        return when (args.size) {
            1 -> SubCommands.entries.map { it.name.lowercase()}
            else -> {listOf()}
        }
    }
}
